File: python-server/src/aerospikeworkshop/routers/retail.py
"""Retail API endpoints — mirrors RetailController.java."""

from __future__ import annotations

import time

from fastapi import APIRouter, Request

router = APIRouter(prefix="/rest/v1", tags=["retail"])


def _product_dict(product) -> dict:
    return product.model_dump(by_alias=True, mode="json")


@router.get("/health")
async def health() -> dict:
    return {
        "status": "OK",
        "message": "FastAPI server is running",
        "timestamp": int(time.time() * 1000),
    }


@router.get("/home")
async def get_home(request: Request) -> dict:
    kv = request.app.state.kv_service
    shoes = await kv.query("subCategory", "Shoes", 10)
    bags = await kv.query("subCategory", "Bags", 10)
    wallets = await kv.query("subCategory", "Wallets", 10)
    watches = await kv.query("subCategory", "Watches", 10)
    headwear = await kv.query("subCategory", "Headwear", 10)
    return {
        "Shoes": [_product_dict(p) for p in shoes.products],
        "Bags": [_product_dict(p) for p in bags.products],
        "Wallets": [_product_dict(p) for p in wallets.products],
        "Watches": [_product_dict(p) for p in watches.products],
        "Headwear": [_product_dict(p) for p in headwear.products],
    }


@router.get("/get")
async def get_product(request: Request, prod: str) -> dict:
    kv = request.app.state.kv_service
    product = await kv.get_product(prod)
    if product is None:
        return {"error": "Product not found"}
    return {
        "error": None,
        "product": _product_dict(product),
        "related": [],
        "also_bought": [],
    }


@router.get("/search")
async def search(
    request: Request,
    q: str | None = None,
    category: str | None = None,
    articleType: str | None = None,
    usage: str | None = None,
    brandName: str | None = None,
) -> dict:
    kv = request.app.state.kv_service
    result = await kv.advanced_search(category, articleType, usage, brandName, q, 20)
    return {
        "products": [_product_dict(p) for p in result.products],
        "time": result.time_ms,
    }


@router.get("/categories")
async def get_categories(request: Request) -> dict:
    categories = await request.app.state.kv_service.get_categories()
    return {"categories": categories, "count": len(categories)}


@router.get("/article-types")
async def get_article_types(request: Request) -> dict:
    article_types = await request.app.state.kv_service.get_article_types()
    return {"articleTypes": article_types, "count": len(article_types)}


@router.get("/usage-types")
async def get_usage_types(request: Request) -> dict:
    usage_types = await request.app.state.kv_service.get_usage()
    return {"usageTypes": usage_types, "count": len(usage_types)}


@router.get("/brand-names")
async def get_brand_names(request: Request) -> dict:
    brand_names = await request.app.state.kv_service.get_brand_names()
    return {"brandNames": brand_names, "count": len(brand_names)}


@router.get("/category")
async def get_category(request: Request, idx: str, filter_value: str) -> dict:
    kv = request.app.state.kv_service
    result = await kv.query(idx, filter_value, 20)
    return {
        "products": [_product_dict(p) for p in result.products],
        "time": result.time_ms,
    }

File: rust-server/src/routers/retail.rs
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

use axum::extract::{Query, State};
use axum::routing::get;
use axum::Json;
use axum::Router;
use serde::Serialize;
use serde_json::{json, Value};

use crate::models::product::Product;
use crate::startup::AppState;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct ProductResponse {
    #[serde(flatten)]
    product: Product,
}

pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        .route("/rest/v1/health", get(health))
        .route("/rest/v1/home", get(get_home))
        .route("/rest/v1/get", get(get_product))
        .route("/rest/v1/search", get(search))
        .route("/rest/v1/categories", get(get_categories))
        .route("/rest/v1/article-types", get(get_article_types))
        .route("/rest/v1/usage-types", get(get_usage_types))
        .route("/rest/v1/brand-names", get(get_brand_names))
        .route("/rest/v1/category", get(get_category))
}

async fn health() -> Json<Value> {
    let timestamp = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0);
    Json(json!({
        "status": "OK",
        "message": "Rust Axum server is running",
        "timestamp": timestamp,
    }))
}

async fn get_home(State(state): State<Arc<AppState>>) -> Result<Json<Value>, String> {
    let kv = &state.kv_service;
    let shoes = kv.query("subCategory", "Shoes", 10).await.map_err(|e| e.to_string())?;
    let bags = kv.query("subCategory", "Bags", 10).await.map_err(|e| e.to_string())?;
    let wallets = kv.query("subCategory", "Wallets", 10).await.map_err(|e| e.to_string())?;
    let watches = kv.query("subCategory", "Watches", 10).await.map_err(|e| e.to_string())?;
    let headwear = kv.query("subCategory", "Headwear", 10).await.map_err(|e| e.to_string())?;

    Ok(Json(json!({
        "Shoes": shoes.products,
        "Bags": bags.products,
        "Wallets": wallets.products,
        "Watches": watches.products,
        "Headwear": headwear.products,
    })))
}

#[derive(serde::Deserialize)]
struct ProductQuery {
    prod: String,
}

async fn get_product(
    State(state): State<Arc<AppState>>,
    Query(params): Query<ProductQuery>,
) -> Result<Json<Value>, String> {
    match state.kv_service.get_product(&params.prod).await {
        Ok(Some(product)) => Ok(Json(json!({
            "error": null,
            "product": product,
            "related": [],
            "also_bought": [],
        }))),
        Ok(None) => Ok(Json(json!({ "error": "Product not found" }))),
        Err(err) => Err(err.to_string()),
    }
}

#[derive(serde::Deserialize)]
struct SearchQuery {
    q: Option<String>,
    category: Option<String>,
    #[serde(rename = "articleType")]
    article_type: Option<String>,
    usage: Option<String>,
    #[serde(rename = "brandName")]
    brand_name: Option<String>,
}

async fn search(
    State(state): State<Arc<AppState>>,
    Query(params): Query<SearchQuery>,
) -> Result<Json<Value>, String> {
    let result = state
        .kv_service
        .advanced_search(
            params.category.as_deref(),
            params.article_type.as_deref(),
            params.usage.as_deref(),
            params.brand_name.as_deref(),
            params.q.as_deref(),
            20,
        )
        .await
        .map_err(|e| e.to_string())?;

    Ok(Json(json!({
        "products": result.products,
        "time": result.time_ms,
    })))
}

async fn get_categories(State(state): State<Arc<AppState>>) -> Result<Json<Value>, String> {
    let categories = state.kv_service.get_categories().await.map_err(|e| e.to_string())?;
    Ok(Json(json!({
        "categories": categories,
        "count": categories.len(),
    })))
}

async fn get_article_types(State(state): State<Arc<AppState>>) -> Result<Json<Value>, String> {
    let values = state.kv_service.get_article_types().await.map_err(|e| e.to_string())?;
    Ok(Json(json!({
        "articleTypes": values,
        "count": values.len(),
    })))
}

async fn get_usage_types(State(state): State<Arc<AppState>>) -> Result<Json<Value>, String> {
    let values = state.kv_service.get_usage().await.map_err(|e| e.to_string())?;
    Ok(Json(json!({
        "usageTypes": values,
        "count": values.len(),
    })))
}

async fn get_brand_names(State(state): State<Arc<AppState>>) -> Result<Json<Value>, String> {
    let values = state.kv_service.get_brand_names().await.map_err(|e| e.to_string())?;
    Ok(Json(json!({
        "brandNames": values,
        "count": values.len(),
    })))
}

#[derive(serde::Deserialize)]
struct CategoryQuery {
    idx: String,
    filter_value: String,
}

async fn get_category(
    State(state): State<Arc<AppState>>,
    Query(params): Query<CategoryQuery>,
) -> Result<Json<Value>, String> {
    let result = state
        .kv_service
        .query(&params.idx, &params.filter_value, 20)
        .await
        .map_err(|e| e.to_string())?;
    Ok(Json(json!({
        "products": result.products,
        "time": result.time_ms,
    })))
}
